WhitespaceAround ignores empty constructor whitespaces

Check documentation: https://checkstyle.org/config_whitespace.html#WhitespaceAround
```
>javac Test.java
>cat Test.java 
package com.example;

public class Test {
    private Test(){} // Does not worn about absent space between ")" and "{"
}

>cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="WhitespaceAround">
            <property name="allowEmptyConstructors" value="true"/>
        </module>
    </module>
</module>
>java $RUN_LOCALE -jar checkstyle-8.32-all.jar -c config.xml Test.java
Starting audit...
Audit done
```
Expected: violation on the 4th line (constructor declaration). There should be space between `){`

Note, that this violation works Ok without `allowEmptyConstructors` option, but an extra violation is reported about empty constructors:
```
>cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="WhitespaceAround">
        </module>
    </module>
</module>

>java $RUN_LOCALE -jar checkstyle-8.32-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /Users/pavel/tmp/17/Test.java:4:19: '{' is not followed by whitespace. [WhitespaceAround]
[ERROR] /Users/pavel/tmp/17/Test.java:4:19: '{' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /Users/pavel/tmp/17/Test.java:4:20: '}' is not preceded with whitespace. [WhitespaceAround]
Audit done.
Checkstyle ends with 3 errors.
```

I tested on Chekcstyle 8.32

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java
@@ -340,7 +340,7 @@ import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
  * </p>
  * <pre>
  * class Test {
- *     public Test(){} // ok
+ *     public Test() {} // ok
  *     public void muFunction() {} // violation, '{' is not followed by whitespace.
  * }
  * </pre>
@@ -767,9 +767,10 @@ public class WhitespaceAroundCheck extends AbstractCheck {
      * @return true if it should be checked if next token is separated by whitespace,
      *      false otherwise.
      */
-    private static boolean shouldCheckSeparationFromNextToken(DetailAST ast, char nextChar) {
-        return !(ast.getType() == TokenTypes.LITERAL_RETURN
-                    && ast.getFirstChild().getType() == TokenTypes.SEMI)
+    private boolean shouldCheckSeparationFromNextToken(DetailAST ast, char nextChar) {
+        return !isEmptyCtorBlockCheckedFromSlist(ast)
+                && !(ast.getType() == TokenTypes.LITERAL_RETURN
+                && ast.getFirstChild().getType() == TokenTypes.SEMI)
                 && ast.getType() != TokenTypes.ARRAY_INIT
                 && !isAnonymousInnerClassEnd(ast.getType(), nextChar)
                 && !isPartOfDoubleBraceInitializerForNextToken(ast);
@@ -799,7 +800,7 @@ public class WhitespaceAroundCheck extends AbstractCheck {
      */
     private boolean isEmptyBlock(DetailAST ast, int parentType) {
         return isEmptyMethodBlock(ast, parentType)
-                || isEmptyCtorBlock(ast, parentType)
+                || isEmptyCtorBlockCheckedFromRcurly(ast)
                 || isEmptyLoop(ast, parentType)
                 || isEmptyLambda(ast, parentType)
                 || isEmptyCatch(ast, parentType);
@@ -888,17 +889,35 @@ public class WhitespaceAroundCheck extends AbstractCheck {
 
     /**
      * Test if the given {@code DetailAST} is part of an allowed empty
-     * constructor (ctor) block.
+     * constructor (ctor) block checked from RCURLY.
      *
      * @param ast the {@code DetailAST} to test.
-     * @param parentType the token type of {@code ast}'s parent.
      * @return {@code true} if {@code ast} makes up part of an
      *         allowed empty constructor block.
      */
-    private boolean isEmptyCtorBlock(DetailAST ast, int parentType) {
+    private boolean isEmptyCtorBlockCheckedFromRcurly(DetailAST ast) {
+        final DetailAST parent = ast.getParent();
+        final DetailAST grandParent = ast.getParent().getParent();
         return allowEmptyConstructors
-                && (isEmptyBlock(ast, parentType, TokenTypes.CTOR_DEF)
-                    || isEmptyBlock(ast, parentType, TokenTypes.COMPACT_CTOR_DEF));
+                && parent.getFirstChild().getType() == TokenTypes.RCURLY
+                && (grandParent.getType() == TokenTypes.CTOR_DEF
+                        || grandParent.getType() == TokenTypes.COMPACT_CTOR_DEF);
+
+    }
+
+    /**
+     * Test if the given {@code DetailAST} is a part of an allowed
+     * empty constructor checked from SLIST token.
+     *
+     * @param ast the {@code DetailAST} to test.
+     * @return {@code true} if {@code ast} makes up part of an
+     *          empty constructor block.
+     */
+    private boolean isEmptyCtorBlockCheckedFromSlist(DetailAST ast) {
+        return allowEmptyConstructors
+                && (ast.getParent().getType() == TokenTypes.CTOR_DEF
+                        || ast.getParent().getType() == TokenTypes.COMPACT_CTOR_DEF)
+                && ast.getFirstChild().getType() == TokenTypes.RCURLY;
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheckTest.java
@@ -511,6 +511,11 @@ public class WhitespaceAroundCheckTest
             "49:14: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "="),
             "49:14: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "="),
             "62:21: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "="),
+            "75:18: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "{"),
+            "79:18: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "{"),
+            "79:18: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "79:40: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "84:23: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "{"),
         };
         verify(checkConfig,
                 getNonCompilablePath("InputWhitespaceAroundAllowEmptyCompactCtors.java"),

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundAllowEmptyCompactCtors.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundAllowEmptyCompactCtors.java
@@ -66,3 +66,20 @@ public class InputWhitespaceAroundAllowEmptyCompactCtors {
         public TestRecord7 {} // ok
     }
 }
+
+class foo {
+    public foo() {}
+}
+
+class foo2 {
+    public foo2(){} // violation, no WS before '{'
+}
+
+class foo3 {
+    public foo3(){System.out.println();} // violation, no WS before and after '{'
+                                    //-^--- violation, no WS before '}'
+}
+
+record TestRecord8(int a, int b) {
+    public TestRecord8{} // violation, no WS before '{'
+}

--- a/src/xdocs/config_whitespace.xml
+++ b/src/xdocs/config_whitespace.xml
@@ -3365,7 +3365,7 @@ class Test {
         <p>Example:</p>
         <source>
 class Test {
-    public Test(){} // ok
+    public Test() {} // ok
     public void muFunction() {} // violation, '{' is not followed by whitespace.
 }
         </source>

